### ISSUE TITLE
Some Text Still Remains, part 2

### ISSUE THREAD

**Issue by MrFanservice-png** [VoxelCubes/PanelCleaner#100]
I set the mask max standard deviation to the 99.99 and some text still remained. What other settings should I change?

2024-07-04 12:36:30.886 | INFO     | pcleaner.gui.launcher:launch:100 - 
- Program Information -
Program: Panel Cleaner 2.7.0
Executing from: C:\Users\<username>\Documents\PanelCleaner\_internal\pcleaner\gui\launcher.pyc
Log file: C:\Users\<username>\AppData\Roaming\pcleaner\cache\pcleaner.log
Config file: C:\Users\<username>\AppData\Roaming\pcleaner\pcleanerconfig.ini
Cache directory: C:\Users\<username>\AppData\Roaming\pcleaner\cache
- System Information -
Operating System: Windows 10
Machine: AMD64
Python Version: 3.11.0 (main, Oct 24 2022, 18:26:48) [MSC v.1933 64 bit (AMD64)]
PySide (Qt) Version: 6.7.0
Available Qt Themes: windows11, windowsvista, Windows, Fusion
System locale: en_US
CPU Cores: 12
GPU: None (CUDA not available)

2024-07-04 12:36:30.980 | INFO     | pcleaner.gui.launcher:launch:114 - Using locale en_US.
2024-07-04 12:36:30.987 | DEBUG    | pcleaner.gui.launcher:launch:121 - Loaded built-in Qt translations for en_US.
2024-07-04 12:36:30.988 | DEBUG    | pcleaner.gui.launcher:launch:129 - Loaded built-in Qt base translations for en_US.
2024-07-04 12:36:31.187 | DEBUG    | pcleaner.gui.mainwindow_driver:ensure_models_downloaded:351 - Text detector model already downloaded.
2024-07-04 12:36:31.189 | DEBUG    | pcleaner.gui.mainwindow_driver:initialize_ui:210 - Purging missing profiles.
2024-07-04 12:36:31.189 | INFO     | pcleaner.gui.mainwindow_driver:initialize_profiles:863 - Found profiles: [('Default', None)]
2024-07-04 12:36:31.189 | DEBUG    | pcleaner.config:load_profile:1152 - Loading profile None...
2024-07-04 12:36:31.189 | DEBUG    | pcleaner.config:load_profile:1159 - Loading builtin default profile
2024-07-04 12:36:31.275 | DEBUG    | pcleaner.gui.mainwindow_driver:load_current_profile:1021 - Loading current profile.
2024-07-04 12:36:31.275 | DEBUG    | pcleaner.gui.profile_parser:set_profile_values:432 - Setting profile values
2024-07-04 12:36:31.279 | DEBUG    | pcleaner.gui.mainwindow_driver:initialize_analytics_view:630 - Loading included font from C:\Users\<username>\Documents\PanelCleaner\_internal\pcleaner\data\NotoMono-Regular.ttf
2024-07-04 12:36:31.280 | DEBUG    | pcleaner.gui.mainwindow_driver:initialize_analytics_view:633 - Loaded included font
2024-07-04 12:36:31.282 | DEBUG    | pcleaner.gui.mainwindow_driver:save_default_palette:129 - Placeholder color: #000000
2024-07-04 12:36:31.287 | INFO     | pcleaner.gui.mainwindow_driver:set_theme:148 - Using system theme.
2024-07-04 12:36:31.287 | INFO     | pcleaner.gui.mainwindow_driver:changeEvent:198 - Theme is dark: False
2024-07-04 12:36:31.289 | INFO     | pcleaner.gui.mainwindow_driver:set_theme:169 - Theme is dark: False
2024-07-04 12:36:31.369 | INFO     | pcleaner.gui.mainwindow_driver:changeEvent:198 - Theme is dark: False
2024-07-04 12:36:31.372 | DEBUG    | pcleaner.gui.mainwindow_driver:post_init:412 - Char width: 6, columns: 74, required width: 444
2024-07-04 12:36:31.374 | DEBUG    | pcleaner.gui.mainwindow_driver:post_init:443 - Splitter sizes: [375, 423, 463]
2024-07-04 12:36:31.375 | DEBUG    | pcleaner.gui.mainwindow_driver:start_initialization_worker:561 - Worker Thread cleaning cache
2024-07-04 12:36:31.375 | DEBUG    | pcleaner.gui.mainwindow_driver:start_initialization_worker:568 - Worker Thread loading OCR model.
2024-07-04 12:36:31.376 | ERROR    | pcleaner.ocr.ocr_tesseract:available_langs:24 - Error checking Tesseract available language data: tesseract is not installed or it's not in your PATH. See README file for more information.
2024-07-04 12:36:31.376 | INFO     | pcleaner.ocr.ocr_mangaocr:__new__:15 - Creating the MangaOcr instance
2024-07-04 12:36:31.377 | INFO     | manga_ocr.ocr:__init__:13 - Loading OCR model from kha-white/manga-ocr-base
2024-07-04 12:36:33.435 | INFO     | manga_ocr.ocr:__init__:25 - Using CPU
2024-07-04 12:36:34.097 | INFO     | manga_ocr.ocr:__init__:32 - OCR ready
2024-07-04 12:36:34.097 | INFO     | pcleaner.gui.mainwindow_driver:load_ocr_model:585 - Loaded OCR model (2.72s)
2024-07-04 12:37:10.256 | DEBUG    | pcleaner.gui.file_table:handleDrop:197 - Dropped C:/Users/<username>/Downloads/Keijo!!!!!!!!/Keijo!!!!!!!! v03 [Akito]/178.png
2024-07-04 12:37:10.256 | DEBUG    | pcleaner.gui.file_table:add_file:246 - Requesting to add "C:\Users\<username>\Downloads\Keijo!!!!!!!!\Keijo!!!!!!!! v03 [Akito]\178.png"
2024-07-04 12:37:10.257 | DEBUG    | pcleaner.gui.file_table:handleDrop:197 - Dropped C:/Users/<username>/Downloads/Keijo!!!!!!!!/Keijo!!!!!!!! v03 [Akito]/187.png
2024-07-04 12:37:10.257 | DEBUG    | pcleaner.gui.file_table:add_file:246 - Requesting to add "C:\Users\<username>\Downloads\Keijo!!!!!!!!\Keijo!!!!!!!! v03 [Akito]\187.png"
2024-07-04 12:37:10.258 | DEBUG    | pcleaner.gui.file_table:handleDrop:197 - Dropped C:/Users/<username>/Downloads/Keijo!!!!!!!!/Keijo!!!!!!!! v03 [Akito]/137_z.png
2024-07-04 12:37:10.258 | DEBUG    | pcleaner.gui.file_table:add_file:246 - Requesting to add "C:\Users\<username>\Downloads\Keijo!!!!!!!!\Keijo!!!!!!!! v03 [Akito]\137_z.png"
2024-07-04 12:37:10.258 | DEBUG    | pcleaner.gui.file_table:repopulate_table:303 - Repopulating table
2024-07-04 12:37:10.263 | INFO     | pcleaner.gui.file_table:lazy_load_images:619 - Dispatching image loading workers
2024-07-04 12:37:10.263 | DEBUG    | pcleaner.gui.file_table:lazy_load_images:625 - Worker Thread loading image C:\Users\<username>\Downloads\Keijo!!!!!!!!\Keijo!!!!!!!! v03 [Akito]\178.png
2024-07-04 12:37:10.263 | DEBUG    | pcleaner.gui.file_table:lazy_load_images:625 - Worker Thread loading image C:\Users\<username>\Downloads\Keijo!!!!!!!!\Keijo!!!!!!!! v03 [Akito]\187.png
2024-07-04 12:37:10.264 | DEBUG    | pcleaner.gui.file_table:lazy_load_images:625 - Worker Thread loading image C:\Users\<username>\Downloads\Keijo!!!!!!!!\Keijo!!!!!!!! v03 [Akito]\137_z.png
2024-07-04 12:37:24.886 | DEBUG    | pcleaner.gui.mainwindow_driver:apply_profile:1130 - Applying profile.
2024-07-04 12:37:33.075 | DEBUG    | pcleaner.gui.mainwindow_driver:save_profile:1160 - User canceled profile save.
2024-07-04 12:37:41.386 | INFO     | pcleaner.gui.mainwindow_driver:save_profile:1173 - Saving profile to C:\Users\<username>\AppData\Roaming\pcleaner\profiles\pp.conf
2024-07-04 12:37:41.386 | DEBUG    | pcleaner.config:unsafe_write:848 - Writing profile to disk...
2024-07-04 12:37:41.389 | DEBUG    | pcleaner.gui.mainwindow_driver:save_profile:1188 - Profile saved successfully.
2024-07-04 12:37:41.412 | WARNING  | pcleaner.gui.mainwindow_driver:profile_change_check:1033 - Profile change check
2024-07-04 12:37:41.413 | DEBUG    | pcleaner.config:load_profile:1152 - Loading profile 'pp'...
2024-07-04 12:37:41.413 | DEBUG    | pcleaner.config:load_profile:1166 - Loading profile pp from C:\Users\<username>\AppData\Roaming\pcleaner\profiles\pp.conf
2024-07-04 12:37:41.413 | INFO     | pcleaner.config:load:862 - Loading profile C:\Users\<username>\AppData\Roaming\pcleaner\profiles\pp.conf from disk...
2024-07-04 12:37:41.415 | DEBUG    | pcleaner.gui.mainwindow_driver:load_current_profile:1021 - Loading current profile.
2024-07-04 12:37:41.415 | DEBUG    | pcleaner.gui.profile_parser:set_profile_values:432 - Setting profile values
2024-07-04 12:37:41.418 | DEBUG    | pcleaner.config:load_profile:1152 - Loading profile 'pp'...
2024-07-04 12:37:41.418 | DEBUG    | pcleaner.config:load_profile:1166 - Loading profile pp from C:\Users\<username>\AppData\Roaming\pcleaner\profiles\pp.conf
2024-07-04 12:37:41.418 | INFO     | pcleaner.config:load:862 - Loading profile C:\Users\<username>\AppData\Roaming\pcleaner\profiles\pp.conf from disk...
2024-07-04 12:37:41.420 | DEBUG    | pcleaner.gui.mainwindow_driver:load_current_profile:1021 - Loading current profile.
2024-07-04 12:37:41.420 | DEBUG    | pcleaner.gui.profile_parser:set_profile_values:432 - Setting profile values
2024-07-04 12:37:43.389 | WARNING  | pcleaner.gui.mainwindow_driver:profile_change_check:1033 - Profile change check
2024-07-04 12:37:43.389 | DEBUG    | pcleaner.config:load_profile:1152 - Loading profile 'Default'...
2024-07-04 12:37:43.389 | DEBUG    | pcleaner.config:load_profile:1159 - Loading builtin default profile
2024-07-04 12:37:43.390 | DEBUG    | pcleaner.gui.mainwindow_driver:load_current_profile:1021 - Loading current profile.
2024-07-04 12:37:43.390 | DEBUG    | pcleaner.gui.profile_parser:set_profile_values:432 - Setting profile values
2024-07-04 12:37:45.039 | WARNING  | pcleaner.gui.mainwindow_driver:profile_change_check:1033 - Profile change check
2024-07-04 12:37:45.040 | DEBUG    | pcleaner.config:load_profile:1152 - Loading profile 'pp'...
2024-07-04 12:37:45.040 | DEBUG    | pcleaner.config:load_profile:1166 - Loading profile pp from C:\Users\<username>\AppData\Roaming\pcleaner\profiles\pp.conf
2024-07-04 12:37:45.040 | INFO     | pcleaner.config:load:862 - Loading profile C:\Users\<username>\AppData\Roaming\pcleaner\profiles\pp.conf from disk...
2024-07-04 12:37:45.042 | DEBUG    | pcleaner.gui.mainwindow_driver:load_current_profile:1021 - Loading current profile.
2024-07-04 12:37:45.042 | DEBUG    | pcleaner.gui.profile_parser:set_profile_values:432 - Setting profile values
2024-07-04 12:37:45.732 | DEBUG    | pcleaner.gui.mainwindow_driver:apply_profile:1130 - Applying profile.
2024-07-04 12:37:54.659 | INFO     | pcleaner.gui.mainwindow_driver:start_cleaning:1317 - Requested outputs: [<Output.denoised_output: 19>, <Output.write_output: 22>]
2024-07-04 12:37:54.662 | INFO     | pcleaner.gui.processing:generate_output:176 - Running text detection AI model for 3 images...
2024-07-04 12:37:56.396 | DEBUG    | pcleaner.ctd_interface:process_image:195 - Saving json file to C:\Users\<username>\AppData\Roaming\pcleaner\cache\cleaner\8b42d6a5-4eea-42bc-b4af-bc6d7ff279d4_178#raw.json
2024-07-04 12:37:56.402 | DEBUG    | pcleaner.image_ops:visualize_raw_boxes:894 - Loading included font from C:\Users\<username>\Documents\PanelCleaner\_internal\pcleaner\data\LiberationSans-Regular.ttf
2024-07-04 12:37:57.867 | DEBUG    | pcleaner.ctd_interface:process_image:195 - Saving json file to C:\Users\<username>\AppData\Roaming\pcleaner\cache\cleaner\a245fe69-9098-4a94-a28b-58a7410a1881_187#raw.json
2024-07-04 12:37:57.873 | DEBUG    | pcleaner.image_ops:visualize_raw_boxes:894 - Loading included font from C:\Users\<username>\Documents\PanelCleaner\_internal\pcleaner\data\LiberationSans-Regular.ttf
2024-07-04 12:37:59.281 | DEBUG    | pcleaner.ctd_interface:process_image:195 - Saving json file to C:\Users\<username>\AppData\Roaming\pcleaner\cache\cleaner\fba2432f-1da4-43de-9a4b-09207622ad5e_137_z#raw.json
2024-07-04 12:37:59.286 | DEBUG    | pcleaner.image_ops:visualize_raw_boxes:894 - Loading included font from C:\Users\<username>\Documents\PanelCleaner\_internal\pcleaner\data\LiberationSans-Regular.ttf
2024-07-04 12:37:59.542 | INFO     | pcleaner.gui.processing:generate_output:228 - Running preprocessing for 3 images...
2024-07-04 12:37:59.542 | DEBUG    | pcleaner.preprocessor:prep_json_file:120 - Processing json file: C:\Users\<username>\AppData\Roaming\pcleaner\cache\cleaner\8b42d6a5-4eea-42bc-b4af-bc6d7ff279d4_178#raw.json
2024-07-04 12:37:59.543 | DEBUG    | pcleaner.preprocessor:prep_json_file:161 - Detected lang: ja
2024-07-04 12:38:00.995 | DEBUG    | pcleaner.preprocessor:prep_json_file:120 - Processing json file: C:\Users\<username>\AppData\Roaming\pcleaner\cache\cleaner\a245fe69-9098-4a94-a28b-58a7410a1881_187#raw.json
2024-07-04 12:38:00.996 | DEBUG    | pcleaner.preprocessor:prep_json_file:161 - Detected lang: ja
2024-07-04 12:38:02.932 | DEBUG    | pcleaner.preprocessor:prep_json_file:120 - Processing json file: C:\Users\<username>\AppData\Roaming\pcleaner\cache\cleaner\fba2432f-1da4-43de-9a4b-09207622ad5e_137_z#raw.json
2024-07-04 12:38:02.933 | DEBUG    | pcleaner.preprocessor:prep_json_file:161 - Detected lang: eng
2024-07-04 12:38:02.936 | INFO     | pcleaner.gui.processing:generate_output:290 - Running masker for 3 images...
2024-07-04 12:38:02.936 | INFO     | pcleaner.gui.mainwindow_driver:show_current_progress:1529 - Showing ocr analytics...
2024-07-04 12:38:10.754 | INFO     | pcleaner.gui.processing:generate_output:404 - Running denoiser for 3 images...
2024-07-04 12:38:10.754 | INFO     | pcleaner.gui.mainwindow_driver:show_current_progress:1538 - Showing masker analytics...
2024-07-04 12:38:17.727 | INFO     | pcleaner.gui.processing:generate_output:602 - Finished processing 3 images.
2024-07-04 12:38:17.727 | INFO     | pcleaner.gui.mainwindow_driver:show_current_progress:1548 - Showing denoiser analytics...
2024-07-04 12:38:17.775 | DEBUG    | pcleaner.image_ops:save_optimized:858 - Saving image 178_clean.png with kwargs: {'optimize': True, 'compress_level': 9, 'dpi': (72.009, 72.009)}
2024-07-04 12:38:18.179 | DEBUG    | pcleaner.image_ops:save_optimized:858 - Saving image 187_clean.png with kwargs: {'optimize': True, 'compress_level': 9, 'dpi': (72.009, 72.009)}
2024-07-04 12:38:18.503 | DEBUG    | pcleaner.image_ops:save_optimized:858 - Saving image 137_z_clean.png with kwargs: {'optimize': True, 'compress_level': 9, 'dpi': (72.009, 72.009)}
2024-07-04 12:38:20.598 | INFO     | pcleaner.gui.mainwindow_driver:output_worker_result:1461 - Output worker finished.
2024-07-04 12:38:44.233 | DEBUG    | pcleaner.gui.mainwindow_driver:open_issue_reporter:760 - Opening issue reporter.

![137_z_clean](https://github.com/VoxelCubes/PanelCleaner/assets/171995610/1167aa2f-5ca9-4d41-98b5-50f0d90227f8)
![178_clean](https://github.com/VoxelCubes/PanelCleaner/assets/171995610/298ebf27-3a4b-4995-8d97-4f257512727d)
![187_clean](https://github.com/VoxelCubes/PanelCleaner/assets/171995610/d4dfed1e-a92d-447d-b0b5-0710b39b7795)

**Comment by VoxelCubes** [comment]
Click on one of the images in the main table to see it's detailed debug outputs. If the area you want cleaned doesn't show up in the raw ai mask and the raw boxes outputs (the 2nd and 3rd outputs in the list) then it can't be helped, the AI didn't manage to detect the text. Nothing can be done. It isn't perfect and like humans also makes mistakes. You will need to clean those parts manually.

**Comment by MrFanservice-png** [comment]
How do I make it detect the missing areas?

**Comment by VoxelCubes** [comment]
You can't. AI isn't perfect. Sometimes it misses a few things. The project https://github.com/dmMaze/comic-text-detector is what panel cleaner relies on to find the text. If it doesn't find the text, which can happen because your computer isn't as smart as you are, then that's it. You will need to manually double check the work. The analytics on the right will list all pages that had failed bubbles, but if they were never detected in the first place, that would catch this either.

tl;dr it can't be fixed, even computers make mistakes. You must double check manually. Panel Cleaner can only do 98% of the work, the remaining 2% must be done by a human.

Maybe it'll manage 99.9% in the future, but not yet.

**Comment by VoxelCubes** [comment]
Ah, I'm so sorry, I forgot to mention the Inpainter. For text that it does recognize but couldn't mask well like in your previous example, instead of raising the max deviation threshold, try enabling inpainting. By default it's off because it can also worsen output quality. But for your material I think it will do a good job. 
The inpainter will handle text that couldn't be masked well enough and use AI to draw what should cover that area instead.

Hope it helps! Thanks for supplying the log and sample images by the way. It's always helpful.

**Comment by MrFanservice-png** [comment]
Where's the inpainting section?

**Comment by MrFanservice-png** [comment]
I used the inpainter and it didn't work. Anything else?
![Screenshot (839)](https://github.com/VoxelCubes/PanelCleaner/assets/171995610/e4a0bf24-4ddc-4848-975e-315b6a3385da)

**Comment by VoxelCubes** [comment]
Well, as long as that's checked and you pressed apply, which it looks like you did, then anything that failed the max deviation threshold will get inpainted (I hope you reset that one to 15 from 99, otherwise inpainting won't get a chance to trigger). If something still wasn't cleaned, that means it just wasn't recognized as text and nothing can be done. That's as good as Panel Cleaner can do it. It it's perfect.
